### PR TITLE
core(tsc): add type checking to computed artifacts

### PR DESCRIPTION
**Pull request by brendankenny**
Important first step to refactoring computed artifacts to be just regular requires. Big PR, but almost all type stuff.

Introduces the lie that the computed artifact `request*()` methods are part of the `Artifacts` type, which isn't usually true except when calling `myaudit.audit(artifacts, auditContext)`. Follow ups will untangle that.

Only real code changes:
- The only time that the `computedArtifacts` object realized it also had all the artifacts on it was in `MainResource`, where `artifacts.URL` was accessed. Instead, `URL` is now passed in as an artifact dependency (as it should be), but that requires `CriticalRequestChains` also needing `URL` passed in. This needed a lot of small test tweaks.
- The lantern simulator needs a settings object passed in, even if it's empty, to keep tsc happy about destructuring. This required some scattered `settings: {}` around.
- the `ComputedArtifact.compute_` method now *has* to be async. It was too annoying dealing with it either being sync (`URL`, `TraceOfTab`, maybe one more?) or async, so just easier this way. Required a good number of additional async/await scattered around, especially in tests.

Pretty much everything else is just adding type annotations or making code a little more explicit for the compiler.

### CODE REVIEW

File: typings/artifacts.d.ts [hunk 1]
```diff
@@ -254,16 +317,32 @@ declare global {
...
+        first: number;
+        speedIndex: number;
+        frames: {
+          getProgress(): number;
+          getTimeStamp(): number;
+          isProgressInterpolated(): number;
+        }[]
+      }
+
+      // TODO(bckenny): all but navigationStart could actually be undefined.
```
Comment (brendankenny) [inline typings/artifacts.d.ts]: these should be updated in a follow up, but it had too many places where we're treating them as always defined. Whether we were correct or not (and we'll get to fix some bugs) remains to be seen.

File: lighthouse-core/gather/computed/computed-artifact.js [hunk 2]
```diff
@@ -3,53 +3,67 @@
...
+   * `compute_` method on classes derived from ComputedArtifact.
+   * @param {FirstParamType<this['compute_']>} requiredArtifacts
+   * @return {Promise<ReturnType<this['compute_']>>}
    */
-  request(artifacts) {
-    if (this._cache.has(artifacts)) {
-      return Promise.resolve(this._cache.get(artifacts));
+  async request(requiredArtifacts) {
+    const computed = this._cache.get(requiredArtifacts);
+    if (computed) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/computed/computed-artifact.js]: kind of a pain that TS can't process `.has` in the same way, that's kind of the point of `.has` :)

File: lighthouse-core/gather/computed/manifest-values.js [hunk 3]
```diff
@@ -23,86 +25,96 @@ class ManifestValues extends ComputedArtifact {
     return ['hasManifest', 'hasParseableManifest'];
   }
 
+  /** @typedef {(val: NonNullable<LH.Artifacts.Manifest['value']>) => boolean} validator */
```
Comment (patrickhulce) [inline lighthouse-core/gather/computed/manifest-values.js]: lowercase typedef name? 😮

File: lighthouse-core/gather/computed/page-dependency-graph.js [hunk 4]
```diff
@@ -163,7 +172,7 @@ class PageDependencyGraphArtifact extends ComputedArtifact {
       let minCandidate = null;
       let minDistance = Infinity;
       // Find the closest request that finished before this CPU task started
-      candidates.forEach(candidate => {
```
Comment (patrickhulce) [inline lighthouse-core/gather/computed/page-dependency-graph.js]: does this really change things type interference wise? or just drive-by perf/style?

File: lighthouse-core/gather/computed/page-dependency-graph.js [hunk 5]
```diff
@@ -173,26 +182,29 @@ class PageDependencyGraphArtifact extends ComputedArtifact {
...
       cpuNode.addDependency(minCandidate);
     }
 
+    /** @type {Map<string, CPUNode>} */
     const timers = new Map();
     for (const node of cpuNodes) {
       for (const evt of node.childEvents) {
         if (!evt.args.data) continue;
 
-        const url = evt.args.data.url;
```
Comment (patrickhulce) [inline lighthouse-core/gather/computed/page-dependency-graph.js]: dataUrl makes me think of a data URI, which other url is this prefix disambiguating?

File: lighthouse-core/audits/seo/is-crawlable.js [hunk 6]
```diff
@@ -77,7 +77,9 @@ class IsCrawlable extends Audit {
    * @return {!AuditResult}
    */
   static audit(artifacts) {
-    return artifacts.requestMainResource(artifacts.devtoolsLogs[Audit.DEFAULT_PASS])
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+
+    return artifacts.requestMainResource({devtoolsLog, URL: artifacts.URL})
```
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/audits/seo/is-crawlable.js]: yeah, I thing simplifying it is a good idea. This PR I guess was mostly about making everything explicit so it'll be easier to refactor :)

File: lighthouse-core/gather/computed/computed-artifact.js (hunk 2 above)
Comment (brendankenny) [reply lighthouse-core/gather/computed/computed-artifact.js]: https://github.com/Microsoft/TypeScript/issues/18781 but they don't sound very enthusiastic :)

File: lighthouse-core/gather/computed/manifest-values.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-core/gather/computed/manifest-values.js]: fixed

File: lighthouse-core/gather/computed/page-dependency-graph.js (hunk 4 above)
Comment (brendankenny) [reply lighthouse-core/gather/computed/page-dependency-graph.js]: yeah, tsc bug (or limitation). `minCadidate` lost its `null | NetworkNode` type after the `forEach()`, and with the if eliminating falsey values it ended up thinking it was `cpuNode.addDependency(never);`

File: lighthouse-core/gather/computed/page-dependency-graph.js (hunk 5 above)
Comment (brendankenny) [reply lighthouse-core/gather/computed/page-dependency-graph.js]: purely style. I was wondering where `url` came from below, then got briefly confused by the various `forEach`es. Changed to `argsUrl`...wdyt? `evtUrl`? Can also revert.

**Comment by brendankenny** [comment]
@patrickhulce you can land #5024 and I'll update this

**Comment by brendankenny** [comment]
updated!

**Comment by brendankenny** [comment]
scoped types and rebased :)
